src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndex.java

```diff
@@ -38,7 +70,51 @@ public abstract void saveIndex(int[] ids, float[][] data, String indexPath, Stri
      *
      * @return size of the hnsw index on the disk in KB.
      */
-    public abstract long getIndexSize();
+    public long getIndexSize() {
+        return this.indexSize;
+    }
+
+    public SpaceType getSpaceType() {
+        return this.spaceType;
+    }
+
+    public void close() {
+        Lock writeLock = readWriteLock.writeLock();
+        writeLock.lock();
+        // Autocloseable documentation recommends making close idempotent. We don't expect to doubly close
+        // but this will help prevent a crash in that situation.
+        if (this.isClosed) {
```

Comment:
you don't need to unlock?

Reply:
Good question. The unlock happens in the finally block, so no.

Reply:
but that will only happen if you are returning inside from try. Hence, if isClosed == true, your unlock will not be called.

Reply:
I believe it will still be reached: https://docs.oracle.com/javase/tutorial/essential/exceptions/finally.html.

Take this sample program:
```
import java.util.*;
import java.io.IOException;
class test {

    public static void main(String[] args)
    {
	try {
		System.out.println("Trying");
		throw new IOException("Test exception");
	} catch (Exception ex) {
		throw new RuntimeException("Test exception: " + ex);
	} finally {
		System.out.println("Reached finally");
	}
    }
}
```

The output is:
```
$ java test
Trying
Reached finally
Exception in thread "main" java.lang.RuntimeException: Test exception: java.io.IOException: Test exception
	at test.main(test.java:11)
```